pass.go

```diff
@@ -0,0 +1,1054 @@
+// request handler.
+type RequestHandler interface {
+	Handle(hi http.Handler) (ho http.Handler)
+}
+
+// RegisterRequestHandler allows an importer to hook the request handler, allows
+// a sort of MITM for incoming requests, RequestHandlers must call f with a Pass
+// instance if they want the request to continue via the main program logic
+func (p *Pass) RegisterRequestHandler(f func(http.Handler) http.Handler) {
+	_, filename, line, _ := runtime.Caller(1)
+	fmt.Printf("Added RequestHandler %s:%d\n----\n", filename, line)
+	p.handler = f(p.handler)
+}
+
+// Pass is a struct holding all required variables for wrappers, instanced using
+// the CreatePass function
+type Pass struct {
+	sv           *serverVariables
+	baseTemplate *template.Template
+	handler      http.Handler
+}
+
+// CreatePass returns a pointer to an instance of Pass
+// for use in calls to Main(Pass) in main() functions of hooking applications
+func CreatePass() *Pass {
+	log.SetFlags(0)
+	p := &Pass{
+		&serverVariables{},
+		nil,
+		nil,
+	}
+
+	configFileName := flag.String("cfg", "pass.json", "A path to the program config file")
+	flag.Parse()
+
+	var err error
+	if p.sv.cfg, err = loadConfig(*configFileName); err != nil {
+		log.Fatalf(err.Error())
+	}
+
+	p.sv.trustedProxies = make([]*net.IPNet, 0)
+	p.sv.trustedProxies, _ = helpers.StringArrayToIPNet(p.sv.cfg.TrustedProxies)
+
+	p.sv.integrityHashes = make(map[string]string)
+	p.sv.integrityHashes["css"] = p.getIntegrityHash("static/css.css")
+	p.sv.integrityHashes["js"] = p.getIntegrityHash("static/js.js")
+
+	p.generateSecurityHeaders()
+
+	p.baseTemplate = template.Must(template.ParseFiles(
+		p.ResolveTemplatePath("layout/base.html"),
+		p.ResolveTemplatePath("layout/dbError.html"),
+	))
+
+	p.handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		rv := new(requestVariables)
+		rv.sv = p.sv
+		rv.w = w
+		rv.r = r
+
+		onBeforeRequest(rv)
+		handler(p, rv)
+	})
+
+	return p
+}
+
+// Main executes the main program, starts the https server, serves clients
+func Main(pass *Pass) {
+	sv := pass.sv
+
+	sv.db = newMustDatabaseConnect(func() (database, error) {
+		return newDatabaseConnectionVariables(sv)
+	})
+	defer sv.db.close()
+
+	preparseTemplates(pass)
+
+	http.Handle("/", recoverWrap(pass.handler))
+
+	_, mainApp, _, _ := runtime.Caller(0)
+	_, wrapper, _, _ := runtime.Caller(1)
+
+	fmt.Printf("%s: Startup!\n", sv.cfg.Title)
+	fmt.Printf("Main program: %s\n", mainApp)
+	fmt.Printf("Wrapped by:   %s\n", wrapper)
+
+	// goroutine that runs forever checking the state of the database connection
+	// This can not run in the normal process flow because of a bug in the pq
+	// module, it is not possible to cancel a query with a timeout
+	// https://github.com/lib/pq/issues/620
+	go func() {
+		for true {
+			if err := sv.db.ping(); err != nil {
+				log.Printf("Database: %v\n", err)
+			}
+
+			time.Sleep(10 * time.Minute)
```

Comment:
Consider using a ticker. https://gobyexample.com/tickers

Reply:
Fixed in next push, assuming `ticker.C` is blocking.